source/vir/src/well_formed.rs

```diff
@@ -94,6 +95,52 @@ fn check_one_expr(
         ExprX::OpenInvariant(_inv, _binder, body, _atomicity) => {
             assert_no_early_exit_in_inv_block(&body.span, body)?;
         }
+        ExprX::AssertQuery { requires, ensures, proof, mode: _ } => {
+            if function.x.attrs.nonlinear {
+                return err_str(
+                    &expr.span,
+                    "assert_by_query not allowed in #[verifier(nonlinear)] functions",
+                );
+            }
+
+            let mut referenced = HashSet::new();
+            for r in requires.iter() {
+                referenced.extend(referenced_vars_expr(r).into_iter());
+            }
+            for r in ensures.iter() {
+                referenced.extend(referenced_vars_expr(r).into_iter());
+            }
+
+            fn check(expr: &Expr, referenced: &HashSet<crate::ast::Ident>) -> Result<(), VirErr> {
+                use crate::visitor::VisitorControlFlow;
+
+                match crate::ast_visitor::expr_visitor_dfs(
+                    expr,
+                    &mut crate::ast_visitor::VisitorScopeMap::new(),
+                    &mut |scope_map, e| match &e.x {
+                        ExprX::Var(x) | ExprX::VarLoc(x) => {
+                            match scope_map.scope_and_index_of_key(&x) {
+                                None if !referenced.contains(x) => VisitorControlFlow::Stop(error(
+                                    format!("variable {} not mentioned in requires", x).as_str(),
```

Comment:
This says "not mentioned in requires", but it looks like `referenced` contains both requires and ensures, not just requires, because of `for r in ensures.iter()` above.  So does the check `check(ens, &referenced)?` below do anything?  If not, you're just checking that `proof` doesn't mention extra variables?  Why does this matter?

Reply:
The ensure check was indeed spurious (leftover from a more restrictive incorrect check I wrote before). The point of this is to reject the following:
```
        #[proof]
        fn test6(x: int, y: int, z:int) {
            requires(y == 0);
            requires(z == 0);
            assert_nonlinear_by({
                requires(y == 0);
                ensures(x + y == x);
                assert(z == 0);
            });
        }
```
as I believe this is unlikely to be what the user intends, as the nonlinear query has no facts about `z`. Maybe this is superfluous?

Reply:
(With the current implementation we wouldn't import typing invariants for variables only mentioned in the body.)